pkg/ccl/crosscluster/logical/logical_replication_job_test.go

```diff
@@ -1003,38 +1003,204 @@ func TestHeartbeatCancel(t *testing.T) {
+	dbA.Exec(t, "UPSERT INTO tab VALUES (1, 'insertA')")
+	dbB.Exec(t, "UPSERT INTO tab VALUES (1, 'insertB')")
+
+	expectedRowsS1 := [][]string{
+		{"1", "insertA"},
+	}
+	expectedRowsDest := [][]string{
+		{"1", "insertB"},
+	}
+	now = s.Clock().Now()
+	WaitUntilReplicatedTime(t, now, dbC, jobAID)
+	WaitUntilReplicatedTime(t, now, dbC, jobBID)
+	dbA.CheckQueryResults(t, "SELECT * from tab", expectedRowsS1)
+	dbB.CheckQueryResults(t, "SELECT * from tab", expectedRowsDest)
+	dbC.CheckQueryResults(t, "SELECT * from tab", expectedRowsDest)
+}
+
+// TestFullyConnectedReplication tests 4 tables that are all streaming
+// from each other and how they handle conflicts
+func TestFullyConnectedReplication(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+
+	clusterArgs := base.TestClusterArgs{
+		ServerArgs: base.TestServerArgs{
+			DefaultTestTenant: base.TestControlsTenantsExplicitly,
+			Knobs: base.TestingKnobs{
+				JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+			},
+		},
+	}
+
+	verifyExpectedRowAllServers := func(
+		t *testing.T, runners []*sqlutils.SQLRunner, expectedRows [][]string, dbNames []string,
+	) {
+		for i, name := range dbNames {
+			runners[i].CheckQueryResults(t, fmt.Sprintf("SELECT * from %s.tab", name), expectedRows)
+		}
+	}
+
+	waitUntilReplicatedTimeAllServers := func(
+		t *testing.T,
+		targetTime hlc.Timestamp,
+		runners []*sqlutils.SQLRunner,
+		jobIDs [][]jobspb.JobID,
+		srcDB int,
+		numDBs int,
+	) {
+		for destDB := range numDBs {
+			if destDB == srcDB {
+				continue
+			}
+			WaitUntilReplicatedTime(t, targetTime, runners[destDB], jobIDs[destDB][srcDB])
+		}
+	}
+
+	numDBs := 4
+	server, s, runners, dbNames := setupServerWithNumDBs(t, ctx, clusterArgs, 1, numDBs)
+	defer server.Stopper().Stop(ctx)
+
+	PGURLs, cleanup := GetPGURLs(t, s, dbNames)
+	defer cleanup()
+
+	// Each row is a DB, each column is a jobID from another DB to that target DB
+	jobIDs := make([][]jobspb.JobID, numDBs)
+	for i := range numDBs {
+		jobIDs[i] = make([]jobspb.JobID, numDBs)
+		for j := range numDBs {
+			if i == j {
+				jobIDs[i][j] = jobspb.InvalidJobID
+				continue
+			}
+			runners[i].QueryRow(t, "CREATE LOGICAL REPLICATION STREAM FROM TABLE tab ON $1 INTO TABLE tab", PGURLs[j].String()).Scan(&jobIDs[i][j])
+		}
+	}
+
+	runners[0].Exec(t, "UPSERT INTO tab VALUES (1, 'celery')")
+	now := s.Clock().Now()
+	waitUntilReplicatedTimeAllServers(t, now, runners, jobIDs, 0, 4)
+
+	expectedRows := [][]string{
+		{"1", "celery"},
+	}
+	verifyExpectedRowAllServers(t, runners, expectedRows, dbNames)
+
+	for i := range numDBs {
+		runners[i].Exec(t, fmt.Sprintf("UPSERT INTO tab VALUES (2, 'row%v')", i))
+	}
+	now = s.Clock().Now()
+	waitUntilReplicatedTimeAllServers(t, now, runners, jobIDs, 0, 4)
+
+	// Depending on a race different rows can win. We care more that they all settled on the same one
```

Comment:
this doesn't solve the bug. row3 should always win. i think there's a bug in `waituntilReplicatedTimeAllServers` -- you're not checking every jobID.

Reply:
in general, i think the signature of this function can be cleaned up.

Reply:
nit: stray docstring.